### ISSUE TITLE
Leading/Middle Rest Elements in Tuple Types with generic function derive unknown

### ISSUE THREAD

**Issue by ashidaharo** [microsoft/TypeScript#47487]
# Bug Report

I originally discovered this when I was tinkering with a function that combines the return values of multiple functions into the argument of the last function given, such as ``reselect``.  However, as a result of verification, it was found that it can be reproduced just by using Leading/Middle Rest Elements in Tuple Types and generic functions at the same time, which is shown in the simple example below.
### 🔎 Search Terms

Leading Middle Rest Tuple Arg Argument generic unknown any

### 🕗 Version & Regression Information

- This is the behavior in every version from 4.2.3, Leading/Middle Rest Elements in Tuple Types implemented, and I reviewed the FAQ, but couldn't find anything related to this.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?jsx=0#code/PTAEBkFMEMBMEsB2BzAsAKBKAypALqAK4AOGeAnsZBDAigEqQDOeAKs25dQLygA8AQQBOyJqEgAPPJESwxhRAGtEAewDuiANoBdADShGTQgBs8APgAUAOhvQRTAFyabV4aP3Wb8aQFtHoNyYASlBuMwNmEzxtELCIo1MAbgxYSABjYztqNJVEFlBjWiRkQzYOADEFNLx4XIcaOGLS9hZWLmT0DCwAORVQAFEhIRUhLrAcvIJCxoYOFrxKxGraxHq+TQAjFRVpxD1QLZ2YREs7UQB9AAZ6w939M6ZzgEZ6i29IPyub7d3Y8Nvjn8Dj9jhgJvlVINhkIAMLQJiQb5HaCIUIFIqzVoVKo1XIWABm0GMCI88CBZI6Y1AbSooAA5AplOpEHTQPAxKoCPCmPBkIhoBtCqA8H0KLS6QCUXSrBYAEwAZllsqCVPBUwxJTm2KWuNW-E2IJR+0Zqg0p3sX2ByMQ9wtL1Ab18j2uVt+oXCJuZQM9GjBuXykCGIzhCKRuzR0zomqxLEWyzxhOJkH0fFYlng9VYQUz7rZQUpnUwYAAsvBYLBClTcAQSGQuKBS+XCs05vXeIJ7OIpDI5EQlKatPtDFF9OVtpYXA8nGOVPoXIEPC53n56oEgcPTDFcxu8B1UhksqA1aAfGWK5AW7GcSt6o3z5fOFQCz0+lCRqr-QRT02L1qrzqb31SU9n0YD9EQExjHNC4XQg4xjFtC57TA0AHnOWVXmXZ0w0BXNgKBYC-UmUAmGgHxIF6N9YXhRFXWONFv3vP8FmvPE4IQ0BExJB0yVzCkMCpGlqAZftmVZdlQE5VCmB5PkBSFEVhXrCVDRZGUFSVFUiyPT8TzPZtmLjXU1gNa1jVEjRwMg6DsMkyDEMeZDVIc9DMKdS18NzH1EG9CzECI-I8AAC0gAB5YLAyokNaOAhj9N-GMWIAtj7M4oluNTdNM2zak+PzATCywbB4FPTIhAOQgCFUcQg3KtIaLEAAqZqWtatr2o6zquo6qtiiFNwq3wIhSHQMVqGKlBCjcaRWjbfg3H0HcJ1sewnAWnjfFXER10iTcdoSXcUnSMrsl0nlJsgabtXjPUJuQKaRBmx9IGfcYzr6y7Huu4z+BQ4CbI85zULtNyPhwlECNUyHrQCggVEUSjavB1FeHO+7PuQJ6jJWAl0uTHjyXywq3uItGHsx77AL4P7VIBl0ULQ+1HTBuiIbwqH2Zh9Bj3h1ggvZaKAEkmER6FkbRMmMax1jEFxpMUzTN5spzOJ+MLQShGgeBjGKQaaxGsbqU17WmmYoS0Q7UQu2kWR5D8nRFt28xPCsKdNBdrCttEfaR1Aed7C3OIdz3Y7D2PPBjZ1zEyn-G7M0j03EqE16dOIiOtaj6MY6SuP9QzVmQILunMOR6G7mBpDS853ZYdAchmFFkZxd4dOTej+ZsbxN5yX0LjICJj804T9vKbqIDVP2f6LDQl1mc+emObiBmQYLsvQW53SVCC7BCCEYghHZYpG6EZvhWHrOO5lixMqV6kctYPje7xgfifiPA9eGutaV6IQfCJB9zbtkCNbHsdsmQaAdvEKIlh8RVFeEuTaAR7A+1MHOFaogHBrm3E7EOB4hCnWIqoX+-9DIy3qD-P+xgAHtAKlgY8RDKEPk7nqdY+dgKT1pgSOBPEq5LyBmwxe-xVK1xUPiGEKhd4ImPqfBhJDErMIsN3MI8An5JhfnQ3SsiqGkOSiw0yuwOHWhgdwuegNrRrxRC5Be5jq7r2PGoIK5AZEjEYTom619Fb5yzCrcIKi0pqILEAA)

### 💻 Code

```ts
// Leading
// Set up
type LeadingRestTestType = <Args extends unknown[], Result>(...args:[...Args, (...items: Args) => Result]) => Result;
declare const leadingRestTestFunction: LeadingRestTestType;

// No Error
// const leadingRestTestFunction: <[boolean], boolean>(args_0: boolean, args_1: (items_0: boolean) => boolean) => boolean
const noErrorCase: boolean = leadingRestTestFunction(false, (i) => i);

// Type 'unknown' is not assignable to type 'boolean'.(2322)
// const leadingRestTestFunction: <[boolean], unknown>(args_0: boolean, args_1: (items_0: boolean) => unknown) => unknown
const errorCase: boolean = leadingRestTestFunction(false, <T>(i: T): T => i);


// Middle
// Set up
type MiddleRestTestType = <Args extends unknown[], Result, Foo>(...args:[Foo, ...Args, (...items: Args) => Result]) => Result;
declare const middleRestTestFunction: MiddleRestTestType;

// No Error
// const middleRestTestFunction: <[boolean], boolean, null>(args_0: null, args_1: boolean, args_2: (items_0: boolean) => boolean) => boolean
const sameNoErrorCase: boolean = middleRestTestFunction(null, false, (i) => i);

// Type 'unknown' is not assignable to type 'boolean'.(2322)
// const middleRestTestFunction: <[boolean], unknown, null>(args_0: null, args_1: boolean, args_2: (items_0: boolean) => unknown) => unknown
const theOtherErrorCase: boolean = middleRestTestFunction(null, false, <T>(i: T): T => i);
```
I also listed other examples in Playground that I suspected would cause the same error because of their similarity, but did not cause the error.

### 🙁 Actual behavior

When I used Leading/Middle Rest Elements in Tuple Types and generic functions together, the result was inferred to be unknown instead of boolean.
This only happened with Leading/Middle Rest Elements in Tuple Types, unlike Traling, Single Arg, Normal Rest Argument.

### 🙂 Expected behavior

At the very least, the Result should be correctly inferred into a boolean.
Also, from the point of view of implementing the function, it should be undesirable to have an error in the case of Leading or Middle, even though it does not occur in the case of Trailing.
It seems that there is a more complicated issue between this feature and generic types, because in type inference, when the type ``<T>(i: T): T => i`` is instantiated, it is inferred as ``args_1: (items_0: boolean) => unknown``.

**Comment by david-shortman** [comment]
Looks related to #47226

**Comment by RyanCavanaugh** [comment]
I can't tell from the report here what you consider expected vs not expected. All of these inferences appear to be sound, and some are underspecific due to limitations in the inference algorithm. If there's a particular case you'd like me to speak to I'm happy to, but I don't have bandwidth to explain what's going on in all 60 lines of code. What's the shortest example you'd consider to be broken?

**Comment by ashidaharo** [comment]
I'm sorry that I didn't try to record my trial and error in order to isolate when the problem **only** occurred.
> some are underspecific due to limitations in the inference algorithm.
And maybe this is one of the issues you are referring to in this way. But just in case, let me rewrite a more correct and concise statement.
> What's the shortest example you'd consider to be broken?
### 💻 Code


```ts
// Leading
// Set up
type LeadingRestTestType = <Args extends unknown[], Result>(...args:[...Args, (...items: Args) => Result]) => Result;
declare const leadingRestTestFunction: LeadingRestTestType;

// No Error
// const leadingRestTestFunction: <[boolean], boolean>(args_0: boolean, args_1: (items_0: boolean) => boolean) => boolean
const noErrorCase: boolean = leadingRestTestFunction(false, (i) => i);

// Type 'unknown' is not assignable to type 'boolean'.(2322)
// const leadingRestTestFunction: <[boolean], unknown>(args_0: boolean, args_1: (items_0: boolean) => unknown) => unknown
const errorCase: boolean = leadingRestTestFunction(false, <T>(i: T): T => i);
```

### 🙁 Actual behavior


In errorCase, Type of return value of leadingRestTestFunction is infered ``unknown``. 

### 🙂 Expected behavior


+ I think this is ``boolean``.
+ and when the type ''<T>(i: T): T => i'' is instantiated, it is inferred as ''args_1: (items_0: boolean) => unknown''. that is a mystery to me This is because the information that the original function had, that the type of the only argument ``i`` and the type of the return value are the same, appears to have been lost.

**Comment by ahejlsberg** [comment]
Issue no longer reproduces, looks like it was fixed in 5.1.
